scripts/ups_build_scripts/ups/v6_0_8/Linux64bit+3.10-2.17/man/man/dropit.1
.TH DROPIT 1 "Fermi"
.SH NAME
dropit \- remove colon separated fields containing substrings
.SH SYNOPSIS
.B "dropit [\-[aefscDEPS]] ... [\-i\fIinput_delimiter\fB] [\-d\fIoutput_delimiter\fB] [\-n\fInull_replacement\fB] [\-p\fIstring\fB] [\fIstring\fB] ..."
.SH DESCRIPTION
\fBdropit\fP's intended purpose is to remove program library names from a
PATH shell variable value, which has colon separated fields.
\fBdropit\fP is usable in sh, ksh, and csh shell script and source files.
\fBdropit\fP  initially always removes the specified field or fields except
when restricted from doing so by the \fB-S\fP option for the specifications of
`/bin' and `.'.  Null fields are replaced by the null field replacement
value, the default is `.' which is the UNIX current directory specification.
This corresponds to the meaning of a null field in a path specification.
.PP
Even though the program's intended usage is adjustment of shell program
PATH environment variable values, the program has been and may be used
as a general purpose delimited character string editor.  The program may
be difficult to use for such purposes a because of its design.
.PP
\fBThe selection options:\fP
.TP 5
.B -e
specifies that the field match is to be exact.
This non-default option is the recommended mode of selection behavior.
.TP
.B -a
specifies that the field match is to be anchored to the beginning of the field.
This non-default option should be used only after due consideration.
.TP
.I none
the default is a substring match anywhere in the field.
Using this default option is most likely not a good idea;
using it frequently causes surprising results.
This was the original behavior of this program and was
implememented this way because of the intended use by
the original requestor.
.PP
\fBThe non-selection options:\fP
.TP
.B -s
specifies that the arguments are to be reinserted into the field list.
By default, those insertions are at the end of field list;
see the \fB-f\fP option for placement at the beginning of the field list.
This combination can be given as \fB-sf\fP.
.TP
.B -f
specifies, if \fB-s\fP is also specified, that the insertions are to
done at the beginning of the field list.
.TP
.B -c
specifies defaults appropriate for the C shell are to be set; otherwise,
defaults for Bourne and Korn shell are set.  The default output delimiter
becomes ` ' instead of `:'.
.TP
.B -D
specifies that duplicate specifications are to be removed retaining the
first occurrence.
.TP
.B -E
specifies that any nonexistent or non-directory entries are to be removed.
.TP
.B -P
specifies that an appropriate list of system directories are to be added
at the front of the field list \fIafter\fP the argument or arguments specified
removals and insertions have been done.
The purpose of this is to protect one from system programs being silently
replaced by non system and probably non-standard variant programs.
.TP
.B -S
specifies a \fIsafer\fP mode of operation is to be used.  A null argument
and an argument of '/bin' can not be removed.  Duplicate removal will be
done even if not otherwise specified.  If the field list is empty after
the removals, insertations and duplicate removals if requested, then
the field list will be made to be an appropriate minimal list of system
directories and checked for being existing directories.  If the field
list remains empty, then it will be made to be `/bin`.  Note, the explicit
current directory specification, `.', can be removed.
.TP
\fB -i\fP delimiter
specifies the character that separates the fields.
.TP
\fB -d\fP delimiter
specifies the string that separates the output fields.
.TP
\fB -n\fP string
specifies null field replacement.  The default value is `.'.
.TP
\fB -p\fP string
specifies the path string to be processed.  If \fB-c\fP is
specified, then the default \fB-i\fP is ` '; otherwise `:'.
.TP
.B [strings] ...
are the fields to be processed.
.SH EXAMPLES
The value of $PATH assumed in the examples below is:
.br
     /usr/sbin:/usr/bsd:/usr/bin:/bin:/etc:/usr/etc:
.br
     /usr/bin/X11:/usr/local/bin://bin

The result of "dropit -d, /etc":
.br
     /usr/sbin,/usr/bsd,/usr/bin,/bin,/usr/bin/X11,
.br
     /usr/local/bin,//bin

The result of "dropit -d' '  /usr //":
.br
     /bin /etc

The value of $path
.br
     /usr/sbin /usr/bsd /usr/bin /bin /etc /usr/etc
.br
     /usr/bin/X11./usr/local/bin //bin

The result of "dropit -i' ' -d' ' -p "$path" /usr //
.br
    /bin /etc

.SH USAGE SUGGESTIONS
Suggested sh and ksh usage in .profile:
.br
     PATH=`dropit -s /etc /funkern`; export PATH

Suggested csh usage in .cshrc or .login:
.br
     set path=(`dropit -sc /etc /funkern`)

Alternative csh usage in .cshrc or .login:
.br
     set path=(`dropit -i' ' -d' ' -p"$path" /etc /funkern`)

Suggested sh and ksh usage in ups setup.sh scripts:
.br
     PATH=`dropit -esf /etc /funkern`; export PATH

Suggested csh usage in ups setup.csh scripts:
.br
     set path=(`dropit -cesf /etc /funkern`)

Alternative csh usage
.br
     set path=(`dropit -i' ' -d' ' -s -f -p"$path" /etc /funkern`)

Suggested sh and ksh usage in ups unsetup.sh scripts:
.br
     PATH=`dropit -e /etc /funkern`; export PATH

Suggested csh usage:
.br
     set path=(`dropit -ce /etc /funkern`)

Alternative csh usage
.br
     set path=(`dropit -i' ' -d' ' -p"$path" /etc /funkern`)

.SH "SEE ALSO"
sh(1), ksh(1) and csh(1)
.SH WARNINGS
A null substring is generally not a good idea;
the results can be very disconcerting.
Recovery may require relogining in, manual specification of a new PATH
environment variable value or manually source'ing the shell's start up scripts.
In the case of C shell variants, the proper way to set the shell PATH
environment variable value is to set the `path' shell non environment
variable value (see also the csh man page on this topic).
.PP
-d' ' and -i' ' indicates a space between two single quotes after the option
letter



scripts/ups_build_scripts/ups/v6_0_8/Linux64bit+3.10-2.17/src/dropit.c
/* dropit.c */

/* Note: this program was deliberately written in a K&R style. */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <pwd.h>
#include <limits.h>

static char *command = "<Unknown>";

static char *safe = 0;

struct Cell_s {
    struct Cell_s *next;
    char *value;
};

static struct Cell_s *pathlist = 0;

static struct Cell_s *
make_Cell(str)
char *str;
{
    struct Cell_s *retval = (struct Cell_s *)malloc(sizeof(struct Cell_s));
    if(retval) {
	if(str) {
	    char *value = (char *)malloc(strlen(str) + 1);
	    if(value) {
		(void) strcpy(value,str);
		retval->next = 0;
		retval->value = value;
	    } else {
		free(retval);
		retval = 0;
	    }
	}
    }
    return retval;
}

static void
destroy_Cell(cell)
struct Cell_s *cell;
{
/* 
** This used to not be ifdefed, and sometimes free-ed
** things that were not malloc-ed. (like the "." string,
** and argv[n], etc.
** besides, we free a bunch of stuff and then exit, so
** why waste the cygles? -- mengel
*/
#ifdef slow_me_down_and_free_things_not_malloced
    if(cell) {
        if(cell->value) free(cell->value); 
	free(cell);
    }
#endif
}

static void
add_at_front(list,cell)
struct Cell_s **list, *cell;
{
    if(list && cell) {
	cell->next = *list;
	*list = cell;
    } else {
	printf("%s\n",safe);
	exit(1);
    }
}

static void
add_at_back(list,cell)
struct Cell_s **list, *cell;
{
    if(list && cell) {
	struct Cell_s *previous, *current;
	for(previous = 0, current = *list;
	    current;
	    previous = current, current = current->next) {
	}
	if(previous) {
	    previous->next = cell;
	} else {
	    *list = cell;
	}
	cell->next = 0;
    } else {
	printf("%s\n",safe);
	exit(1);
    }
    
}

static int
contains(field,test)
char *field, *test;
{
    char *place;
    int len = strlen(test);

    if(len == 0) {
	return 1;
    }
    while(place = strchr(field,test[0])) {
	if(strncmp(place,test,len) == 0) {
	    return 1;
	}
	field = place + 1;
    }
    return 0;
}

static int
anchored(field,test)
char *field, *test;
{
    int len = strlen(test);
    if(len == 0) {
	return 1;
    }
    if (strncmp(field,test,len) == 0) {
	return 1;
    }
    return 0;
}

static int
exact(field,test)
char *field, *test;
{
    if(strcmp(field,test) == 0) {
	return 1;
    }
    return 0;
}

int
exp_stat(name, psb) 
char *name; struct stat *psb;
{
   static char replbuf[PATH_MAX];
   struct passwd *pw;
   char *psl;
   /* try to expand ~/ or ~user/ */
   if (name && name[0] == '~') {
       psl = strchr(name, '/');
       if (psl == name+1 || (psl == 0 && name[1] ==0)) {
           strcpy(replbuf, getenv("HOME"));
           if (psl) {
               strcat(replbuf, psl);
           }
           name = replbuf;
       } else {
           if (psl) {
              *psl = 0;
           }
           pw = getpwnam(name+1);
           if( pw && pw->pw_dir ) {
               strcpy(replbuf, pw->pw_dir);
               if( psl ) {
                  *psl = '/';
                   strcat(replbuf, psl);
               }
               name = replbuf;
           }
       }
   }
   return stat(name,psb);
}

main(argc,argv,envp)
int argc;
char **argv;
char **envp;
{
    extern char *getenv();
    int (*compare)() = 0;
    void (*insert)() = 0;
    char *path;
    char *syspath;
    char *new;
    char *odel;
    char *null;
    char idel;
    char **cpp;
    char *temp;
    extern int getopt();
    extern char *optarg;
    extern int optind, opterr;
    int any = 0;
    int opt;
    int error = 0;
    int length;
    int Lodel;
    extern int errno;
    extern char *sys_errlis[];
    int Anchored = 0;
    int Exact = 0;
    int first = 0;
    int cshmode = 0;
    int setup = 0;
    int Safe = 0;
    int duplicates = 1;
    int protected = 0;
    int existance = 0;
    int Ddefault = 1;
    int Edefault = 1;
    int Pdefault = 1;
    int Sdefault = 1;
    int adefault = 1;
    int cdefault = 1;
    int edefault = 1;
    int fdefault = 1;
    int idefault = 1;
    int ndefault = 1;
    int odefault = 1;
    int pdefault = 1;
    int sdefault = 1;
    struct Cell_s *previous, *current, *next;
    struct stat sb;

    if(argv && *argv) {
	char *cp = strrchr(argv[0],'/');
	command = cp ? (cp+1) : *argv;
    }

    while((opt = getopt(argc,argv,"i:d:n:p:aefsciDEPS?")) != -1) {
	switch(opt) {
	case 'D':
	    duplicates = 0;
	    if(Ddefault) Ddefault = 0; else error = 1;
	    break;
	case 'E':
	    existance = 1;
	    if(Edefault) Edefault = 0; else error = 1;
	    break;
	case 'P':
	    protected = 1;
	    if(Pdefault) Pdefault = 0; else error = 1;
	    break;
	case 'S':
	    Safe = 1;
	    if(Sdefault) Sdefault = 0; else error = 1;
	    break;
	case 'a':
	    Anchored = 1;
	    if(adefault) adefault = 0; else error = 1;
	    break;
	case 'e':
	    Exact = 1;
	    if(edefault) edefault = 0; else error = 1;
	    break;
	case 'f':
	    first = 1;
	    if(fdefault) fdefault = 0; else error = 1;
	    break;
	case 'p':
	    path = optarg;
	    if(pdefault) pdefault = 0; else error = 1;
	    break;
	case 'i':
	    idel = optarg[0];
	    if(idefault) idefault = 0; else error = 1;
	    break;
	case 'd':
	    odel = optarg;
	    if(odefault) odefault = 0; else error = 1;
	    break;
	case 'n':
	    null = optarg;
	    if(ndefault) ndefault = 0; else error = 1;
	    break;
	case 's':
	    setup = 1;
	    if(sdefault) sdefault = 0; else error = 1;
	    break;
	case 'c':
	    cshmode = 1;
	    if(cdefault) cdefault = 0; else error = 1;
	    break;
	default:
	    error = 1;
	}
    }

    if(cshmode) {
	if(idefault) idel = pdefault ? ':' : ' ';
	if(odefault) odel = " ";
    } else {
	if(idefault) idel = ':';
	if(odefault) odel = ":";
    }

    /* Create a `safe' path, preferably their current PATH */

    {   char *syspath = getenv("PATH");
	if(!syspath) syspath = "/bin:/sbin:/usr/bin:/usr/bin";

	if(strlen(odel) == 1) {
	    safe = (char *)malloc(strlen(syspath)+1);
	    if(safe) {
		char *cp;
		(void) strcpy(safe,syspath);
		for(cp = safe; *cp; ++cp) if(*cp == ':') *cp = odel[0];
	    }
	} else {
	    char *cp;
	    int del = 0;
	    for(cp = syspath; *cp; ++cp) if(*cp == ':') ++del;
	    safe = (char *)malloc(strlen(syspath)+del*(strlen(odel)-1)+1);
	    if(safe) {
		char byte[2];
		byte[1] = '\0';
		safe[0] = '\0';
		for(cp = syspath; *cp; ++cp) {
		    byte[0] = *cp;
		    if(*cp == ':') {
			strcat(safe,odel);
		    } else {
			strcat(safe,byte);
		    }
		}
	    }
	}
	if(!safe) {
	    if(strlen(odel) == 1) {
		if(odel[0] == ' ') {
		    safe = "/bin /sbin /usr/bin /usr/bin";
		} else if(odel[0] == ':') {
		    safe = "/bin:/sbin:/usr/bin:/usr/bin";
		} else {
		    safe = "/bin";
		}
	    } else {
		safe = "/bin";
	    }
	}
    }
    
    /* After this point, `safe' must not be changed! */

    if(error) {
	fprintf(stderr,
"%s: usage: %s [-[aefscDEPS]] ... [-i x] [-d x] [-p path] [string] ...\n",
	    command,command);
	printf("%s\n",safe);
	exit(1);
    }
    if(pdefault) ndefault = 1;
    if(ndefault) null = ".";

    if(pdefault) {
	path = getenv("PATH");
	if(!path) {
	    idel = ':';
	    path = "/bin:/sbin:/usr/bin:/usr/bin";
	}
    }

    if(path[0]) {
	char *lcp, *rcp;
	char *copy = (char *)malloc(strlen(path)+1);
	if(copy) {
	    (void) strcpy(copy,path);
	} else {
	    fprintf(stderr,"No memory to copy path\n");
	    printf("%s\n",safe);
	    exit(1);
	}
	for(lcp = copy, rcp = lcp; *rcp; ++rcp) {
	    if(*rcp == idel) {
		*rcp = '\0';
		if(*lcp) {
		    add_at_back(&pathlist,make_Cell(lcp));
		} else {
		    add_at_back(&pathlist,make_Cell(null));
		}
		lcp = rcp + 1;
	    }
	}
	if(*lcp) {
	    add_at_back(&pathlist,make_Cell(lcp));
	} else {
	    add_at_back(&pathlist,make_Cell(null));
	}
	free(copy);
    }

    compare = (Exact ? exact : (Anchored ? anchored : contains));
    insert = (first ? add_at_front : add_at_back);

    /* Do removals */
    for(cpp = argv+optind; *cpp; ++cpp) {
	struct Cell_s *next;
	if(Safe) {
	    if(**cpp == '\0') continue;
	    if(strcmp(*cpp,"/bin") == 0) continue;
	}
	for(previous = 0, current = pathlist; current; current = next) {
	    next = current->next;
	    if(compare(current->value,*cpp)) {
		if(previous) {
		    previous->next = current->next;
		} else {
		    pathlist = current->next;
		}
		destroy_Cell(current);
	    } else {
		previous = current;
	    }
	}
    }

    /* Do insertations, if requested */
    if(setup) {
	for(cpp = argv+optind; *cpp; ++cpp) {
	    insert(&pathlist,
		make_Cell((**cpp == '\0') ? null : *cpp));
	}
    }

    /* Do insertations of system director(y)(ies), if requested */
    if(protected) {
	add_at_front(&pathlist,make_Cell("/usr/bin/X11"));
	add_at_front(&pathlist,make_Cell("/usr/etc"));
	add_at_front(&pathlist,make_Cell("/etc"));
	add_at_front(&pathlist,make_Cell("/usr/sbin"));
	add_at_front(&pathlist,make_Cell("/usr/bin"));
	add_at_front(&pathlist,make_Cell("/sbin"));
	add_at_front(&pathlist,make_Cell("/bin"));
	add_at_back(&pathlist,make_Cell("/usr/local/bin"));
    }

    /* Remove duplicates */
    if(Safe || protected || !duplicates) {
	for(current = pathlist; current; current = current -> next) {
	    struct Cell_s *prev, *curr, *next;
	    for(prev = current, curr = current->next; curr; curr = next) {
		next = curr->next;
		if(strcmp(current->value,curr->value) == 0) {
		    prev->next = curr->next;
		    destroy_Cell(curr);
		} else {
		    prev = curr;
		}
	    }
	}
    }

    /* Remove nonexistant file system objects or non-directories, */
    /* if requested */
    if(existance) {
	struct Cell_s *next;
	for(previous = 0, current = pathlist;
	    current;
	    current = current -> next) {
	    next = current->next;
	    sb.st_mode = -1;
	    if(exp_stat(current->value,&sb) ||
	       (sb.st_mode & S_IFMT) != S_IFDIR) {
		if(previous) {
		    previous->next = current->next;
		} else {
		    pathlist = current->next;
		}
		destroy_Cell(current); 
	    } else {
		previous = current;
	    }
	}
    }

    /* Do insertations of minimal system director(y)(ies), */
    /* if requested and needed */
    if(Safe && !pathlist) {
	add_at_front(&pathlist,make_Cell("/usr/sbin"));
	add_at_front(&pathlist,make_Cell("/usr/bin"));
	add_at_front(&pathlist,make_Cell("/sbin"));
	add_at_front(&pathlist,make_Cell("/bin"));
	/* Remove nonexistant file system objects or non-directories */
	for(previous = 0, current = pathlist;
	    current;
	    current = current -> next) {
	    next = current->next;
	    sb.st_mode = -1;
	    if(stat(current->value,&sb) ||
	       (sb.st_mode & S_IFMT) != S_IFDIR) {
		if(previous) {
		    previous->next = current->next;
		} else {
		    pathlist = current->next;
		}
		destroy_Cell(current);
	    } else {
		previous = current;
	    }
	}
	/* Recheck */
	if(!pathlist) add_at_front(&pathlist,make_Cell("/bin"));
    }

    /* Generate the new PATH value */
    for(current = pathlist; current; current = next) {
	next = current->next;
	if(any) printf("%s",odel); else any = 1;
	printf("%s",current->value);
	destroy_Cell(current);
    }
    pathlist = 0;
    printf("\n");

    exit(0);
}

